### PR TITLE
Gradebook

### PR DESCRIPTION
**Pull request by flying-pi**
Add gradebook tab

### CODE REVIEW

File: rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js [hunk 1]
```diff
@@ -565,10 +565,85 @@
...
+
+            function onError() {
+                alert("Can not load statistic fo select course");
+            }
+
+            $.ajax({
+                traditional: true,
+                type: "POST",
+                url: "api/gradebook/",
+                data: {filter:filterString},
```
Comment (idegtiarov) [inline rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js]: nit: I would add trailing space after the colon

File: rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js [hunk 2]
```diff
@@ -565,10 +565,85 @@
...
+            greadebookTab.gradebookTableBody.empty();
+
+            for (var i = 0; i < greadebookTab.examNames.length; i++) {
+                greadebookTab.gradebookTableHeader.append('<th><div class="assignment-label">' + greadebookTab.examNames[i] + '</div></th>')
+            }
+
+            greadebookTab.studentsTable.empty();
+            for (var i = 0; i < greadebookTab.studentInfo.length; i++) {
+                greadebookTab.studentsTable.append(
+                    '<tr class="" style="display: table-row;" >' +
```
Comment (idegtiarov) [inline rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js]: odd `class`

File: rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html [hunk 3]
```diff
@@ -5,3 +5,32 @@
 %>
 
 <h1>${_("Gradebook")}</h1>
+
+<table class="student-table" style="float: left; width: 20%">
```
Comment (idegtiarov) [inline rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html]: I would use css file for handling styles

File: rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html [hunk 4]
```diff
@@ -5,3 +5,32 @@
...
+            </form>
+        </th>
+    </tr>
+    </thead>
+    <tbody id="student_table_body">
+
+    </tbody>
+</table>
+
+<div class="grades" style="position: relative;float: left;width: 30%;overflow-x: auto;overflow-y: hidden">
```
Comment (idegtiarov) [inline rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html]: ditto

File: rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html [hunk 5]
```diff
@@ -5,3 +5,32 @@
...
+    <table class="grade-table">
+        <thead id="gradebook_table_header"/>
+        <tbody id="gradebook_table_body"/>
+    </table>
+</div>
+
+
+<div style="position: relative;float: left;width: 50%;overflow-x: auto;overflow-y: hidden">
```
Comment (idegtiarov) [inline rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html]: ditto

File: rg_instructor_analytics/urls.py [hunk 6]
```diff
@@ -4,11 +4,13 @@
...
 from rg_instructor_analytics.views import (
-    EnrollmentStatisticView, InstructorAnalyticsFragmentView, ProblemDetailView, ProblemHomeWorkStatisticView,
-    ProblemQuestionView, ProblemsStatisticView)
+    EnrollmentStatisticView, GradebookView, InstructorAnalyticsFragmentView, ProblemDetailView,
+    ProblemHomeWorkStatisticView, ProblemQuestionView, ProblemsStatisticView
+)
 
 urlpatterns = [
-    url(r'^api/enroll_statics/$', EnrollmentStatisticView.as_view(), name='enrollment_statistic_view'),
+    url(r'^api/enroll_statics/$', EnrollmentStatisticView.as_view(),
```
Comment (idegtiarov) [inline rg_instructor_analytics/urls.py]: please use OpenEdx Python style guide

File: rg_instructor_analytics/views.py [hunk 7]
```diff
@@ -39,7 +43,44 @@
 QUESTUIN_MULTI_SELECT_TYPE = 'multySelect'
 
 
-class EnrollmentStatisticView(View):
+class AccessMixin(object):
+    """Base class for provide check  user permission."""
```
Comment (idegtiarov) [inline rg_instructor_analytics/views.py]: ditto, here and everywhere

File: rg_instructor_analytics/views.py [hunk 8]
```diff
@@ -39,7 +43,44 @@
...
+    def process(self, request, **kwargs):
+        """Process allowed request."""
+        pass
+
+    def base_process(self, request, course_id):
+        """Preprocess request, check permission and select course."""
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            log.error("Unable to find course with course key %s while loading the Instructor Analytics Dashboard.",
```
Comment (idegtiarov) [inline rg_instructor_analytics/views.py]: ditto

File: rg_instructor_analytics/views.py [hunk 9]
```diff
@@ -396,26 +389,63 @@ def post(self, request, course_id):
...
 
 
-class InstructorAnalyticsFragmentView(EdxFragmentView):
+class GradebookView(AccessMixin, View):
+    """Api for question statistic."""
+
+    def get_grades_values(self, grade_info):
+        """Return percent value of the student grade."""
+        result = [int(g['percent'] * 100.0) for g in grade_info['section_breakdown']]
+        result += [int(grade_info['percent'] * 100.0)]
```
Comment (idegtiarov) [inline rg_instructor_analytics/views.py]: I would use append method for adding one element to the list

File: rg_instructor_analytics/views.py [hunk 10]
```diff
@@ -396,26 +389,63 @@ def post(self, request, course_id):
...
+                for student in enrolled_students
+            ]
+            exam_names = []
+            if len(enrolled_students) > 0:
+                exam_names = [
+                    g['label']
+                    for g in CourseGradeFactory().create(enrolled_students[0], kwargs['course'])
+                                                 .summary['section_breakdown']
+                ]
+                exam_names += [_('total')]
```
Comment (idegtiarov) [inline rg_instructor_analytics/views.py]: ditto

**Review by idegtiarov — COMMENTED** [review]
Some small nits and comments are inline. Other LGTM

File: rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js (hunk 1 above)
Comment (flying-pi) [reply rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js]: done

File: rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js (hunk 2 above)
Comment (flying-pi) [reply rg_instructor_analytics/static/rg_instructor_analytics/js/instructor_analytics.js]: Remove inline style and add set class

File: rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html (hunk 3 above)
Comment (flying-pi) [reply rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html]: done

File: rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html (hunk 4 above)
Comment (flying-pi) [reply rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html]: done

File: rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html (hunk 5 above)
Comment (flying-pi) [reply rg_instructor_analytics/templates/rg_instructor_analytics/gradebook.html]: done

File: rg_instructor_analytics/urls.py (hunk 6 above)
Comment (flying-pi) [reply rg_instructor_analytics/urls.py]: fix

File: rg_instructor_analytics/views.py (hunk 7 above)
Comment (flying-pi) [reply rg_instructor_analytics/views.py]: fix

File: rg_instructor_analytics/views.py (hunk 9 above)
Comment (flying-pi) [reply rg_instructor_analytics/views.py]: changed

File: rg_instructor_analytics/views.py (hunk 10 above)
Comment (flying-pi) [reply rg_instructor_analytics/views.py]: changed

**Review by idegtiarov — APPROVED** [review]

**Pull request merged**